### PR TITLE
core, server: set size limits on userfields

### PR DESCRIPTION
**Pull request by ad-tra**
context: core, server
server: 
- [x] usernames 
- [x] filenames 

core: 
- [x] usernames 
- [x] filenames 

fixes: #1228

### CODE REVIEW

File: libs/core/libs/shared/src/api.rs [hunk 1]
```diff
@@ -93,6 +96,7 @@ pub enum ChangeDocError {
     DocumentNotFound,
     DocumentDeleted,
     NotPermissioned,
+    FileNameTooLong,
```
Comment (ad-tra) [inline libs/core/libs/shared/src/api.rs:99]: > @parth: prob doesn't need to be here
the file name length should be checked both in upsert and when the user renames. So it should be included.

File: libs/core/src/service/account_service.rs [hunk 2]
```diff
@@ -14,6 +14,10 @@ impl<Client: Requester> RequestContext<'_, '_, Client> {
     ) -> CoreResult<Account> {
         let username = String::from(username).to_lowercase();
 
+        if username.len() > 254 {
```
Comment (ad-tra) [inline libs/core/src/service/account_service.rs:17]: > @Parth this can prob be a smaller value lol, I think like 20

**Review by Parth — CHANGES_REQUESTED** [review]
Your approach for usernames is fine. 

Your approach for filenames is not quite what we want:

You don't want to hunt down all the places that names could change. 

Checkout `validate.rs`, specifically the `validate()` function. You can add a size check (on the encrypted copy). And it will apply everywhere (renames, paths, creates etc).

**Review by Parth — CHANGES_REQUESTED** [review]
This looks better, but a few more things required:
+ [x] `get_encryption_overhead` shouldn't be called for usernames, they're always cleartext.
+ [x] that function is only ever called with 1 value (right?) you can just have a const for the `MAX_ENCRYPTED_FILENAME_LENGTH` with the math hardcoded
 + [x] you want to go through `lib.rs` and see which endpoints could cause a filename to become too long (create, rename, etc) and you don't want this new error (`FileNameTooLong`) to be mapped to `Unexpected`. You want it to be handled by our various clients.
 + [ ] when you do this, a handful of clients will break, you want to chase down those compiler errors and CI failures and fix them.
 + [ ] Currently if you use CLI to create a file with a name too long you'll just get an `Unexpected`, the target state would provide a better error message, and our tooling should help you do that. @SmailBarkouch can do iOS and android.
+ [ ] is a UUID 50 characters long? I thought it was shorter than that by a good margin.

File: libs/core/libs/shared/src/validate.rs [hunk 3]
```diff
@@ -80,6 +81,11 @@ where
     // note: deleted access keys permissible
     pub fn assert_all_files_decryptable(&mut self, owner: Owner) -> SharedResult<()> {
         for file in self.ids().into_iter().filter_map(|id| self.maybe_find(id)) {
+            if file.secret_name().encrypted_value.value.len() > MAX_ENCRYPTED_FILENAME_LENGTH {
```
Comment (Parth) [inline libs/core/libs/shared/src/validate.rs:84]: deserves it's own fn, idk if it makes sense to tack this check onto this fn

File: libs/core/tests/file_ops.rs [hunk 4]
```diff
@@ -55,6 +56,13 @@ fn name_empty() {
     assert_matches!(core.rename_file(id, ""), Err(UiError(RenameFileError::NewNameEmpty)));
 }
 
+#[test]
+fn name_invalid() {
+    let core = test_core_with_account();
+    let result = core.create_at_path(&"x".repeat(MAX_FILENAME_LENGTH + 1));
```
Comment (Parth) [inline libs/core/tests/file_ops.rs:62]: is this a copy paste error?

**Review by Parth — CHANGES_REQUESTED** [review]
get this rebased and we'll get the errors piped through this week

**Comment by ad-tra** [comment]
Make sure not to return unexpected

**Comment by Parth** [comment]
Before we merge this, we need to run the new validations with the current server data to make sure there aren't any files currently that are too long

**Review by Parth — APPROVED** [review]

File: libs/core/libs/shared/src/validate.rs [hunk 5]
```diff
@@ -76,7 +78,6 @@ where
 
         Ok(())
     }
-
```
Comment (steverusso) [inline libs/core/libs/shared/src/validate.rs:79]: Leave this in, and then...

File: libs/core/libs/shared/src/validate.rs [hunk 6]
```diff
@@ -92,6 +93,16 @@ where
...
+    pub fn assert_all_filenames_size_limit(&self) -> SharedResult<()> {
+        for file in self.all_files()? {
+            if file.secret_name().encrypted_value.value.len() > MAX_ENCRYPTED_FILENAME_LENGTH {
+                return Err(SharedErrorKind::ValidationFailure(
+                    ValidationFailure::FileNameTooLong(*file.id()),
+                ))?;
+            }
+        }
+        Ok(())
+    }
```
Comment (steverusso) [inline libs/core/libs/shared/src/validate.rs:105]: ... add one to give functions some breathing room.

File: libs/core/src/lib.rs
Comment (steverusso) [inline libs/core/src/lib.rs:1]: @Parth should `create_at_path` also have `FileNameTooLong` as an expected error?

File: server/server/src/utils.rs [hunk 7]
```diff
@@ -1,11 +1,15 @@
 use crate::ServerError;
-use lockbook_shared::api::{GetBuildInfoError, GetBuildInfoResponse};
+use lockbook_shared::{
+    account::MAX_USERNAME_LENGTH,
+    api::{GetBuildInfoError, GetBuildInfoResponse},
+};
```
Comment (steverusso) [inline server/server/src/utils.rs:5]:
```suggestion
use lockbook_shared::account::MAX_USERNAME_LENGTH;
use lockbook_shared::api::{GetBuildInfoError, GetBuildInfoResponse};
```

File: server/server/src/file_service.rs [hunk 8]
```diff
@@ -88,6 +88,7 @@ pub async fn upsert_file_metadata(
 
     for update in request.updates {
         let new = update.new;
+
```
Comment (steverusso) [inline server/server/src/file_service.rs:91]:
```suggestion
```
Try to remove all standalone inconsequential changes like this, esp before final review.

**Review by steverusso — CHANGES_REQUESTED** [review]

File: libs/core/src/lib.rs
Comment (steverusso) [reply libs/core/src/lib.rs:1]: @ad-tra @Parth will `create_at_path` ever return a `FileNameTooLong`? Don't we have to check the path components?

Comment (Parth) [reply libs/core/src/lib.rs:1]: @ad-tra `create_link_at_path` too. Can you throw some sanity-check-tests in there too?

**Review by Parth — CHANGES_REQUESTED** [review]

**Review by steverusso — APPROVED** [review]

**Pull request merged**